File: backend/routing/download/history_get.go
package routes_download

import (
	"aura/cache"
	"aura/database"
	"aura/logging"
	"aura/utils/httpx"
	"net/http"
	"strconv"
)

type GetDownloadHistory_Response struct {
	Entries []database.DownloadHistoryEntry `json:"entries"`
	Total   int                             `json:"total"`
}

// GetDownloadHistory godoc
// @Summary      Download History - Get Entries
// @Description  Retrieve past download history entries (one per poster set per run), including success/warning/error status and per-image failure details.
// @Tags         Download
// @Accept       json
// @Produce      json
// @Param        limit   query     int  false  "Max entries to return (default 50)"
// @Param        offset  query     int  false  "Offset for pagination (default 0)"
// @Security     SessionCookie
// @Security     ApiKeyAuth
// @Failure      401  {object}  httpx.UnauthorizedResponse "Unauthorized (only when Auth.Enabled=true)"
// @Success      200  {object}  httpx.JSONResponse{data=GetDownloadHistory_Response}
// @Failure      500           {object}  httpx.JSONResponse "Internal Server Error"
// @Router       /api/download/history [get]
func GetDownloadHistory(w http.ResponseWriter, r *http.Request) {
	ctx, ld := logging.CreateLoggingContext(r.Context(), r.URL.Path)
	logAction := ld.AddAction("Download History - Get Entries", logging.LevelInfo)
	ctx = logging.WithCurrentAction(ctx, logAction)
	var response GetDownloadHistory_Response

	limit := 50
	if v, err := strconv.Atoi(r.URL.Query().Get("limit")); err == nil && v > 0 {
		limit = v
	}
	offset := 0
	if v, err := strconv.Atoi(r.URL.Query().Get("offset")); err == nil && v >= 0 {
		offset = v
	}

	entries, total, Err := database.GetDownloadHistory(ctx, limit, offset)
	if Err.Message != "" {
		httpx.SendResponse(w, ld, response)
		return
	}

	for i := range entries {
		if ratingKey, found := cache.LibraryStore.GetRatingKeyByTMDBID(entries[i].LibraryTitle, entries[i].TMDB_ID); found {
			entries[i].RatingKey = ratingKey
		}
	}

	response.Entries = entries
	response.Total = total
	httpx.SendResponse(w, ld, response)
}

File: frontend/src/services/downloads/history-get.ts
import apiClient from "@/services/api-client";
import { ReturnErrorMessage } from "@/services/api-error-return";

import { log } from "@/lib/logger";

import type { APIResponse } from "@/types/api/api-response";
import type { DownloadHistoryEntry } from "@/types/database/download-queue";

export interface GetDownloadHistory_Response {
  entries: DownloadHistoryEntry[];
  total: number;
}

export const GetDownloadHistory = async (limit = 50, offset = 0): Promise<APIResponse<GetDownloadHistory_Response>> => {
  try {
    log("INFO", "API - Download History", "Fetch", "Fetching download history entries");
    const response = await apiClient.get<APIResponse<GetDownloadHistory_Response>>(`/download/history`, {
      params: { limit, offset },
    });
    if (response.data.status === "error") {
      throw new Error(response.data.error?.message || "Unknown error fetching download history");
    }
    return response.data;
  } catch (error) {
    log("ERROR", "API - Download History", "Fetch", "Error fetching download history", { error });
    return ReturnErrorMessage<GetDownloadHistory_Response>(error);
  }
};
